test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTest.cs

```diff
@@ -0,0 +1,188 @@
+            var json = JsonConvert.DeserializeObject<Dictionary<string, string>>(body);
+            Assert.Equal(1, json.Count);
+            Assert.Equal("The ProductDetails field is required.", json["product.ProductDetails"]);
+        }
+
+        [Fact]
+        public async Task ModelMetaDataTypeAttribute_InvalidClassAttributeOnBaseClass_ModelStateErrors()
+        {
+            // Arrange
+            var server = TestServer.Create(_services, _app);
+            var client = server.CreateClient();
+            var input = "{ \"Contact\":\"4258959019\", \"Category\":\"Technology\"," +
+                "\"CompanyName\":\"Microsoft\", \"Country\":\"UK\",\"Price\": 21, \"ProductDetails\": {\"Detail1\": \"d1\"," +
+                " \"Detail2\": \"d2\", \"Detail3\": \"d3\"}}";
+
+            var content = new StringContent(input, Encoding.UTF8, "application/json");
+
+            var url = "http://localhost/ModelMetadataTypeValidation/ValidateProductViewModelIncludingMetadata";
+
+            // Act
+            var response = await client.PostAsync(url, content);
+
+            // Assert
+            var body = await response.Content.ReadAsStringAsync();
+            var json = JsonConvert.DeserializeObject<Dictionary<string, string>>(body);
+            Assert.Equal(1, json.Count);
+            Assert.Equal("Product must be made in the USA if it is not named.", json["product"]);
+        }
+
+        [Fact]
+        public async Task ModelMetaDataTypeAttribute_ValidDerivedClass_EmptyResponseBody()
+        {
+            // Arrange
+            var server = TestServer.Create(_services, _app);
+            var client = server.CreateClient();
+            var input = "{ \"Name\": \"MVC\", \"Contact\":\"4258959019\", \"Category\":\"Technology\"," +
+                "\"CompanyName\":\"Microsoft\", \"Country\":\"USA\", \"Version\":\"2\"," +
+                "\"DatePurchased\": \"/Date(1297246301973)/\", \"Price\" : \"110\" }";
+            var content = new StringContent(input, Encoding.UTF8, "application/json");
+
+            var url = "http://localhost/ModelMetadataTypeValidation/ValidateSoftwareViewModelIncludingMetadata";
+
+            // Act
+            var response = await client.PostAsync(url, content);
+
+            // Assert
+            var body = await response.Content.ReadAsStringAsync();
+            Assert.Equal("{}", body);
+        }
+
+        [Fact]
+        public async Task ModelMetaDataTypeAttribute_InvalidPropertiesOnDerivedClass_ModelStateErrors()
+        {
+            // Arrange
+            var server = TestServer.Create(_services, _app);
+            var client = server.CreateClient();
+
+            var url = "http://localhost/ModelMetadataTypeValidation/ValidateSoftwareViewModelIncludingMetadata";
+            var request = new HttpRequestMessage(HttpMethod.Post, url);
+
+            var version = "2.2";
+            request.Headers.TryAddWithoutValidation("version", version);
```

Comment:
not clear how this header affects the test.  if it's not necessary, change the test to use `PostAsync()` and otherwise be more similar to other tests.  if it's required, add code comments explaining the need.